Zend/zend_compile.c

```diff
@@ -4498,13 +4499,26 @@ void zend_compile_return(zend_ast *ast) /* {{{ */
 	if (!expr_ast) {
 		expr_node.op_type = IS_CONST;
 		ZVAL_NULL(&expr_node.u.constant);
-	} else if (by_ref && zend_is_variable(expr_ast)) {
-		zend_compile_var(&expr_node, expr_ast, BP_VAR_W, 1);
 	} else {
-		zend_compile_expr(&expr_node, expr_ast);
+		if (op_array->scope && !(op_array->fn_flags & ZEND_ACC_HAS_RETURN_TYPE)) {
+			if (zend_is_constructor(op_array->function_name)) {
+				zend_error(E_DEPRECATED,
+					"Returning non-void value from a constructor is deprecated");
+			} else if (zend_string_equals_literal_ci(op_array->function_name,
+				ZEND_DESTRUCTOR_FUNC_NAME)) {
+				zend_error(E_DEPRECATED,
+					"Returning now-void value from a destructor is deprecated");
```

Comment:
```suggestion
					"Returning non-void value from a destructor is deprecated");
```

I like this message! Maybe the current method name could be added to the beginning to improve it.

Reply:
> Maybe the current method name could be added to the beginning for an improvement

Sure, will do :)

Reply:
Isn't that "non-void" superfluous, because a value is never void?

Reply:
> Isn't that "non-void" superfluous, because a value is never void?

Makes sense, I'll drop the "non-void" part.

Reply:
Done.